### PR TITLE
API v1

### PR DESCRIPTION
**Pull request by Swordlash**
APIv1 reached the moment that it can be pushed to master. Future work will involve basic neural network implementation using low-level API and APIv2, featuring high level API for networks

### CODE REVIEW

File: API/Tensorflow/native_libs/src/graph/GraphSession.h [hunk 1]
```diff
@@ -52,15 +48,31 @@ class GraphSession
 	}
 
 	template<TF_DataType DataTypeLabel>
-	Tensor<DataTypeLabel>** eval() const
+	Tensor<DataTypeLabel>** eval(const std::map<std::string, std::shared_ptr<Tensor<DataTypeLabel>>>& substitutions) const
```
Comment (radeusgd) [inline API/Tensorflow/native_libs/src/graph/GraphSession.h:51]: This is ok for now, but I think in the future we may want to also have an `eval :: List Placeholder -> List Tensor -> List Tensor` that doesn't need to use placeholder names (just using their pointers is the suggested way by TF C API).
For example, lift may not have to use the strings but just pointer values and get rid of using the map.

Comment (Swordlash) [reply API/Tensorflow/native_libs/src/graph/GraphSession.h:51]: Yes, it is doable.

File: API/Tensorflow/native_libs/src/ops/Const.h [hunk 2]
```diff
@@ -16,7 +16,7 @@ template <TF_DataType DataTypeLabel>
 class Const : public Operation<DataTypeLabel> {
 public:
     explicit Const(std::shared_ptr<Tensor<DataTypeLabel>> tensor) : value(std::move(tensor)) {
-    	hash = (value->hash());
+    	hash = (value->hash() ^ const_salter.next_salt());
```
Comment (radeusgd) [inline API/Tensorflow/native_libs/src/ops/Const.h]: This requires a comment documenting why we decide to do that.

**Comment by radeusgd** [comment]
Full documentation may be postponed until the API is more stable, but before merge at least adding function signatures in the comments for each public function is a minimum.

**Review by radeusgd — CHANGES_REQUESTED** [review]
Full documentation may be postponed until the API is more stable, but before merge at least adding function signatures in the comments for each public function is a minimum.

File: API/Tensorflow/native_libs/src/ops/Const.h (hunk 2 above)
Comment (Swordlash) [reply API/Tensorflow/native_libs/src/ops/Const.h]: Although mathematically two consts sharing the same value are the same objects, we decided to distinguish them. Look at the following piece of code, from the regression example:
```
[a,b] = train . fold [0.0, 0.0] gradientDescent
```
where the step function creates constTensor form those zeros, calculates gradient of error function and calls batchEval. We fold over a list of values, using stochastic gradient descent, and our accumulator is initially a zero scalar. But this way, the step function would treat both of the inputs, wrapped into Const Op, **as the same object**, although we did not intend to treat them that way. That would result in returning **always the same values** for **a** and **b**. If we changed the code into something like
```
[a,b] = train . fold [0.0, 0.01] gradientDescent,
```
it would lead to totally different (and much better) results (e.g, returned a and b would be different, and proper)! Generally, it is not a wise way to differentiate by Const; the more mathematical way is to use Placeholder, or unimplemented yet Variable instead. But because it **is** possible in the API, we have to distinguish consts having the same values but being different objects in some way, and xoring its hash with some uniformly-distributed salt is a way to achieve this.

**Review by radeusgd — APPROVED** [review]
